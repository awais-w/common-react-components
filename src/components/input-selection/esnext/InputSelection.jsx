import React from 'react';
import PropTypes from 'prop-types';
import bem from '@argos/utils/esnext/bem';
import './InputSelection.scss';

const propTypes = {
  checked: PropTypes.bool.isRequired,
  children: PropTypes.any,
  checkboxIconType: PropTypes.oneOf(['tick', 'cross', 'line']),
  id: PropTypes.string.isRequired,
  inputType: PropTypes.oneOf(['checkbox', 'radio']).isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
};

const defaultProps = {
  checkboxIconType: 'tick',
};

const InputSelection = (props) => {
  const { checked, children, checkboxIconType, id, inputType, name, onChange, value } = props;
  const cn = bem({ prefix: 'cmc-', block: 'input-selection' });
  return (
    <div className={cn()}>
      <input
        checked={checked}
        className={cn('input', inputType === 'checkbox' ? checkboxIconType : '')}
        id={id}
        name={name}
        onChange={onChange}
        type={inputType}
        value={value}
      />
      <label htmlFor={id} className={cn('label')}>
        <div className={cn('text-container')}>{children}</div>
      </label>
    </div>
  );
};

InputSelection.defaultProps = defaultProps;
InputSelection.propTypes = propTypes;
export default InputSelection;
