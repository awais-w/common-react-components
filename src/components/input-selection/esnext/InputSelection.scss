@import '@cmc/common-styles/src/scss/_variables.scss';

.cmc-input-selection {
  position: relative;
  display: block;

  label {
    @include font-standard;
    color: $color-grey-charcoal;
    cursor: pointer;
    margin: 0;
    padding: 0;
    display: flex;
    align-items: center;

    :focus {
      outline: inherit;
      outline-offset: $half;
    }
  }

  .cmc-input-selection__text-container {
    padding-left: $base;
  }

  input[type='checkbox'],
  input[type='radio'] {
    margin-right: $base;
    opacity: 0;
    position: absolute;
    align-self: center;

    & + label:before {
      content: '';
      @include prefix(transition, all 0.3s ease, webkit moz ms o);
      font-family: 'Material Icons';
      color: $color-white;
      display: block;
      width: $input-height-regular * 0.5;
      height: $input-height-regular * 0.5;
      background-color: $color-white;
      border: 1px solid $color-grey-dark;
      display: flex;
      align-items: center;
      justify-content: center;
      border-radius: $small;
      box-sizing: border-box;
      @include touch-device-bp1024 {
        width: $large;
        height: $large;
      }
    }

    &:active + label:before {
      box-shadow: 0;
    }

    &:checked.cmc-input-selection__input--tick + label:before {
      content: 'check';
      background-color: $primary-color;
      border-color: $primary-color;
    }

    &:checked.cmc-input-selection__input--cross + label:before {
      content: 'close';
      background-color: $color-rust;
      border-color: $color-rust;
    }

    &:checked.cmc-input-selection__input--line + label:before {
      content: 'remove';
      background-color: $color-sainsburys;
      border-color: $color-sainsburys;
    }

    :focus + label:before {
      border-color: $color-grey-dark;
    }
  }

  input[type='radio'] {
    & + label:before {
      border-radius: 50%;
      content: '';
    }
  }

  input[type='radio']:checked {
    & + label:before {
      border: $half + 2 solid $primary-color;
      @include touch-device-bp1024 {
        border-width: $base;
      }
    }
  }

  input input:focus + label {
    transition: none 0s ease 0s;
  }

  fieldset[disabled] label,
  fieldset[disabled] input,
  input.disabled,
  input[disabled] {
    cursor: not-allowed;
  }
}
