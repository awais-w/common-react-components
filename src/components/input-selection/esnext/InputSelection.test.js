import React from 'react';
import { shallow } from 'enzyme';
import InputSelection from './InputSelection';

const defaultProps = {
  checked: true,
  children: 'Label',
  checkboxIconType: 'tick',
  id: 'checkbox-id',
  inputType: 'checkbox',
  name: 'checkbox-name',
  onChange: jest.fn(),
  value: 'something',
};

const setup = (props) => {
  const component = shallow(<InputSelection {...props} />);

  return {
    wrapper: component,
    getCheckbox: () => component.find('.cmc-input-selection'),
    getInput: () => component.find('.cmc-input-selection__input'),
    getLabel: () => component.find('.cmc-input-selection__label'),
  };
};

describe('<InputSelection />', () => {
  const component = setup(defaultProps);

  it('should match default snapshot', () => {
    expect(component.wrapper).toMatchSnapshot();
  });

  describe('rendering', () => {
    it('exists', () => {
      expect(component.getCheckbox().exists()).toBeTruthy();
    });

    it('renders the input', () => {
      expect(component.getInput().exists()).toBe(true);
    });

    it('renders the label', () => {
      expect(component.getLabel().exists()).toBe(true);
    });

    it('should render the children', () => {
      expect(component.getLabel().children()).toHaveLength(1);
    });
  });

  describe('functionality', () => {
    it('is renders an input of type checkbox', () => {
      expect(component.getInput().props().type).toBe('checkbox');
    });

    it('is checked when the checked prop is true', () => {
      expect(component.getInput().props().checked).toBe(true);
    });

    it('is not checked when the checked prop is false', () => {
      const uncheckedComponent = setup({ ...defaultProps, checked: false });
      expect(uncheckedComponent.getInput().props().checked).toBe(false);
    });

    it('calls onChange when the value changes', () => {
      component.getInput().simulate('change');
      expect(defaultProps.onChange).toHaveBeenCalled();
    });

    it('contains the classname tick when passsed icon tick', () => {
      expect(component.getInput().prop('className')).toContain('tick');
    });

    it('contains the classname cross when passsed icon cross', () => {
      const crossProps = { ...defaultProps, checkboxIconType: 'cross' };
      const crossComponent = setup(crossProps);
      expect(crossComponent.getInput().prop('className')).toContain('cross');
    });

    it('contains the classname cross when passsed icon line', () => {
      const lineProps = { ...defaultProps, checkboxIconType: 'line' };
      const crossComponent = setup(lineProps);
      expect(crossComponent.getInput().prop('className')).toContain('line');
    });

    it('is renders an input of type radio', () => {
      const radioProps = { ...defaultProps, inputType: 'radio' };
      const radioComponent = setup(radioProps);
      expect(radioComponent.getInput().props().type).toBe('radio');
    });
  });
});
