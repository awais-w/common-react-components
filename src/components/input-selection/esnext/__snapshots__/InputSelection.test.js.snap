// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputSelection /> should match default snapshot 1`] = `
<div
  className="cmc-input-selection"
>
  <input
    checked={true}
    className="cmc-input-selection__input cmc-input-selection__input--tick"
    id="checkbox-id"
    name="checkbox-name"
    onChange={[MockFunction]}
    type="checkbox"
    value="something"
  />
  <label
    className="cmc-input-selection__label"
    htmlFor="checkbox-id"
  >
    <div
      className="cmc-input-selection__text-container"
    >
      Label
    </div>
  </label>
</div>
`;
